app/scripts/metamask-controller.test.js

```diff
@@ -2840,4 +2840,144 @@ describe('MetaMaskController', () => {
       expect(browserPolyfillMock.storage.session.set).not.toHaveBeenCalled();
     });
   });
+
+  describe('MetaMetrics & authentication dependencies', () => {
+    let metamaskController;
+    let mockPerformSignIn;
+    let mockPerformSignOut;
+
+    const arrangeControllerState = (stateOverrides) => {
+      metamaskController = new MetaMaskController({
+        showUserConfirmation: noop,
+        encryptor: mockEncryptor,
+        initState: { ...cloneDeep(firstTimeState), ...stateOverrides },
+        initLangCode: 'en_US',
+        platform: {
+          showTransactionNotification: () => undefined,
+          getVersion: () => 'foo',
+        },
+        browser: browserPolyfillMock,
+        infuraProjectId: 'foo',
+        isFirstMetaMaskControllerSetup: true,
+      });
+
+      mockPerformSignIn = jest
+        .spyOn(metamaskController.authenticationController, 'performSignIn')
+        .mockResolvedValue();
+      mockPerformSignOut = jest
+        .spyOn(metamaskController.authenticationController, 'performSignOut')
+        .mockResolvedValue();
+    };
+
+    it('should sign in the user if MetaMetrics is enabled and the user is not signed in', async () => {
+      arrangeControllerState({
+        MetaMetricsController: {
+          participateInMetaMetrics: false,
+        },
+        AuthenticationController: {
+          isSignedIn: false,
+        },
+      });
+
+      metamaskController.controllerMessenger.publish(
+        'MetaMetricsController:stateChange',
+        {
+          participateInMetaMetrics: true,
+        },
+      );
+
+      expect(mockPerformSignIn).toHaveBeenCalledTimes(1);
+      expect(mockPerformSignOut).not.toHaveBeenCalled();
+    });
+
+    it('should sign out the user if MetaMetrics is disabled and the user is signed in but profile syncing is disabled', async () => {
+      arrangeControllerState({
+        MetaMetricsController: {
+          participateInMetaMetrics: true,
+        },
+        AuthenticationController: {
+          isSignedIn: true,
+        },
+        UserStorageController: {
+          isProfileSyncingEnabled: false,
+        },
+      });
+
+      metamaskController.controllerMessenger.publish(
+        'MetaMetricsController:stateChange',
+        {
+          participateInMetaMetrics: false,
+        },
+      );
+
+      expect(mockPerformSignIn).not.toHaveBeenCalled();
+      expect(mockPerformSignOut).toHaveBeenCalledTimes(1);
+    });
+
+    it('should not sign in the user if MetaMetrics is enabled and the user is already signed in', async () => {
+      arrangeControllerState({
```

Comment:
I really appreciate these tests! I was wondering if we could clean em up using a test matrix + `it.each`, but I think the setup required to run the test would be a bit messy. These are fine as is.